File: src/main/java/board/ChessGameInterface.java
package board;

import java.util.List;

import helper.*;
/**
 * @author stephan
 */
public interface ChessGameInterface extends BasicChessGameInterface
{
	final public static int NO_END                    = 0;
	final public static int MATT                      = 1;
	final public static int PATT                      = 2;
	final public static int DRAW                      = 3;
	final public static int THREE_TIMES_SAME_POSITION = 4;
	final public static int FIFTY_MOVES_NO_HIT        = 5;
	final public static int RESIGN                    = 6;

	public void    initGame(  int chess960  );
	public List<ChessGameInterface> copyGame(int neededInstances);
	
	public boolean isWhiteTurn();
	public int     countFigures();
	public boolean hasHitFigure();
	public boolean isCheck( boolean isWhite );
	public boolean isSelectable( Position pos,boolean whitePlayer );
	public boolean isMoveable( Position from,Position to,boolean whitePlayer );
	public boolean isFreeArea(Position pos);
	public int     move(Move move);
	public void getPossibleMoves(List<Move> possibleMoves );
	public int     countReachableMoves( boolean isWhite );

//	public void useNormalSupervisor();
	public ChessGameSupervisor suspendInteractiveSupervisor();
	public void useSupervisor( ChessGameSupervisor supervisor );
	public void undo();
}

File: module-engine-internal/src/main/kotlin/voidchess/engine/board/ChessGameInterface.kt
package voidchess.engine.board

import voidchess.common.board.BasicChessBoard
import voidchess.common.board.other.StartConfig
import voidchess.common.board.move.Move
import voidchess.common.board.move.MoveResult
import voidchess.common.board.move.Position
import voidchess.common.board.other.ChessGameSupervisor


internal interface ChessGameInterface : BasicChessBoard {

    /**
     * returns true if this game was started with the standard distribution of figures
     * (chess960 code: 518)
     */
    val startConfig: StartConfig
    val isWhiteTurn: Boolean
    val hasHitFigure: Boolean
    val isCheck: Boolean

    /**
     * @return all the moves played so far in a string representation
     */
    fun getCompleteHistory(): String
    fun initGame(chess960: Int)
    fun copyGame(neededInstances: Int): List<ChessGameInterface>
    fun countFigures(): Int
    fun isSelectable(pos: Position, whitePlayer: Boolean): Boolean
    fun isMovable(from: Position, to: Position, whitePlayer: Boolean): Boolean
    fun move(move: Move): MoveResult
    fun getAllMoves(): Collection<Move>
    fun getCriticalMoves(): Collection<Move>
    fun getTakingMoves(): Collection<Move>
    /**
     * @return a pair of ints with the first int denoting the reachable sum of moves of all white figures
     * while the second int is the same for all black figures.
     */
    fun countReachableMoves(): Pair<Int, Int>
    fun suspendInteractiveSupervisor(): ChessGameSupervisor
    fun useSupervisor(supervisor: ChessGameSupervisor)
    fun undo()
}

internal fun <T> ChessGameInterface.withMove(move: Move, workWithGameAfterMove: (ChessGameInterface)->T): T {
    move(move)
    return workWithGameAfterMove(this).apply {
        undo()
    }
}